src/nlp.jl

```diff
@@ -90,147 +98,307 @@ struct NLPBlockData
 end
 
 """
-    initialize(d::AbstractNLPEvaluator, requested_features::Vector{Symbol})
+    initialize(
+        d::AbstractNLPEvaluator,
+        requested_features::Vector{Symbol},
+    )::Nothing
+
+Initialize `d` with the set of features in `requested_features`. Check
+[`features_available`](@ref) before calling `initialize` to see what features
+are supported by `d`.
+
+!!! warning
+    This method must be called before any other methods.
+
+## Features
+
+The following features are defined:
+
+ * `:Grad`: enables [`eval_objective_gradient`](@ref)
+ * `:Jac`: enables [`eval_constraint_jacobian`](@ref)
+ * `:JacVec`: enables [`eval_constraint_jacobian_product`](@ref) and
+   [`eval_constraint_jacobian_transpose_product`](@ref)
+ * `:Hess`: enables [`eval_hessian_lagrangian`](@ref)
+ * `:HessVec`: enables [`eval_hessian_lagrangian_product`](@ref)
+ * `:ExprGraph`: enables [`objective_expr`](@ref) and [`constraint_expr`](@ref).
+
+In all cases, including when `requested_features` is empty,
+[`eval_objective`](@ref) and [`eval_constraint`](@ref) are supported.
 
-Must be called before any other methods. The vector `requested_features`
-lists features requested by the solver. These may include `:Grad` for gradients
-of the obejctive, ``f``, `:Jac` for explicit Jacobians of constraints, ``g``,
-`:JacVec` for Jacobian-vector products, `:HessVec` for Hessian-vector
-and Hessian-of-Lagrangian-vector products, `:Hess` for explicit Hessians and
-Hessian-of-Lagrangians, and `:ExprGraph` for expression graphs.
+## Examples
+
+```julia
+MOI.initialize(d, Symbol[])
+MOI.initialize(d, [:ExprGraph])
+MOI.initialize(d, MOI.features_available(d))
+```
 """
 function initialize end
 
 """
-    features_available(d::AbstractNLPEvaluator)
+    features_available(d::AbstractNLPEvaluator)::Vector{Symbol}
+
+Returns the subset of features available for this problem instance.
 
-Returns the subset of features available for this problem instance, as a
-vector of symbols in the same format as in `initialize`.
+See [`initialize`](@ref) for the list of defined features.
 """
 function features_available end
 
 """
-    eval_objective(d::AbstractNLPEvaluator, x)
+    eval_objective(d::AbstractNLPEvaluator, x::AbstractVector{Float64})::Float64
```

Comment:
Should we assume that we are always working with `Float64`? I would prefer being more generic there, if possible. For instance:

```julia
eval_objective(d::AbstractNLPEvaluator, x::AbstractVector{T})::T where T 

```

Reply:
JuliaSmoothOptimizer has solvers supporting `Float32`, and they developed a bridge between NLPModels.jl and MOI. But you are right: if everything is hard-coded to use `Float64` in `Nonlinear`, it would be misleading to change the documentation to be type-independent

Reply:
Okay, I've removed `Float64` from the docstrings. But there's a lot of other `Float64`s hard-coded into `nlp.jl` and `MOI.Nonlinear`, so I don't think we have a viable way of making them generic. There's nothing stopping the creation of a different `AbstractNLPEvaluator` that is generic though.